### PR TITLE
Send tracer events to cli in test mode

### PR DESCRIPTION
**Pull request by adamnolte**
Synchronously sends events to the cli when we detect a cli environment variable. The cli just collects these into an array, so doing this synchronously will have minimal impact on the test run.

Uses context variables to associate test id and test case hash. The test is setup so that it verifies when we have multiple test cases running at a time, everything is working correctly.

### CODE REVIEW

File: autoblocks/_impl/testing/run.py [hunk 1]
```diff
@@ -141,7 +145,7 @@ async def run_test_case_unsafe(
             testCaseOutput=serialize(output),
         ),
     )
-
+    current_test_case_hash.reset(token)
```
Comment (adamnolte) [inline autoblocks/_impl/testing/run.py:148]: this isn't necessary, but more of a safety mechanism

File: autoblocks/_impl/testing/run.py [hunk 2]
```diff
@@ -256,6 +261,7 @@ async def async_run_test_suite(
         )
 
     await global_state.http_client().post(f"{cli()}/end", json=dict(testExternalId=test_id))
+    current_external_test_id.reset(token)
```
Comment (adamnolte) [inline autoblocks/_impl/testing/run.py:264]: this isn't necessary, but more of a safety mechanism

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -52,6 +52,7 @@ jobs:
 
   tests:
     runs-on: ubuntu-latest
+    timeout-minutes: 5
```
Comment (adamnolte) [inline .github/workflows/ci.yml:55]: i had an issue with freezegun and async tests so added this to prevent ci from just running forever when the tests hang

**Comment by adamnolte** [comment]
I did also test this in a python project (user feedback clustering)

File: autoblocks/_impl/tracer.py [hunk 4]
```diff
@@ -244,6 +231,45 @@ async def _send_event_unsafe(
...
+            log.error("Failed to send test event to Autoblocks. No test case hash set.")
+            return
+
+        traced_event = TracerEvent(
+            message=message,
+            trace_id=trace_id,
+            timestamp=timestamp,
+            properties=properties,
+        )
+        req = global_state.sync_http_client().post(
```
Comment (nicolewhite) [inline autoblocks/_impl/tracer.py:262]: Would prefer we just use the async client for everything to be consistent but I guess this works if we're doing it on the main thread and not in our event loop

File: autoblocks/_impl/tracer.py [hunk 5]
```diff
@@ -264,16 +290,38 @@ def send_event(
         the event failed, the trace_id will be None.
         """
         try:
+            merged_properties = dict(self._properties)
```
Comment (nicolewhite) [inline autoblocks/_impl/tracer.py:293]: Would be nice to pull all of this property merging stuff into a function just bc I think functions with a large amount of lines in a try/except are hard to follow

**Review by nicolewhite — DISMISSED** [review]
Left some thoughts but not blockers

File: autoblocks/_impl/testing/run.py (hunk 2 above)
Comment (nicolewhite) [reply autoblocks/_impl/testing/run.py:264]: Do we want to reset to the previous value? Or just set it to None

Comment (adamnolte) [reply autoblocks/_impl/testing/run.py:264]: The previous value would be none here

File: autoblocks/_impl/tracer.py (hunk 4 above)
Comment (nicolewhite) [reply autoblocks/_impl/tracer.py:262]: Oh wait I guess that wouldn't even work right? It was blocking?

File: autoblocks/_impl/testing/run.py (hunk 2 above)
Comment (nicolewhite) [reply autoblocks/_impl/testing/run.py:264]: Oh right ok

File: autoblocks/_impl/tracer.py (hunk 4 above)
Comment (adamnolte) [reply autoblocks/_impl/tracer.py:262]: Well this function is running in our event loop since it's in the test context. I think we'll always want this to be blocking in the test context

Comment (nicolewhite) [reply autoblocks/_impl/tracer.py:262]: Oh well in that case it would be nice to use the async client because when you use a sync client you're blocking the loop until it completes instead of yielding to other tasks

Comment (adamnolte) [reply autoblocks/_impl/tracer.py:262]: Since it's just sending to the local cli server (and that just adds to an in memory array and returns) it should be very fast.

I'd have to use event-loop.execute to use async right? I can see if it can get that work

Comment (nicolewhite) [reply autoblocks/_impl/tracer.py:262]: I think in the testing context it doesn't matter too much what we do here. And I think we're going to switch to queue + sync anyway right? So maybe we can leave it

Comment (adamnolte) [reply autoblocks/_impl/tracer.py:262]: Ya I was just opting for simple in test context. I'm not sure if we'll want queue for test events. We'd have to make sure to flush it before we send test case results

Comment (nicolewhite) [reply autoblocks/_impl/tracer.py:262]: Ya I meant for prod

File: autoblocks/_impl/testing/run.py (hunk 1 above)
Comment (nicolewhite) [reply autoblocks/_impl/testing/run.py:148]: it's also possible this wouldn't be reached if there is an uncaught error in `fn`, you'd have to put this in a `finally` block

File: autoblocks/_impl/tracer.py [hunk 6]
```diff
@@ -244,6 +231,66 @@ async def _send_event_unsafe(
...
+            json=dict(
+                testExternalId=current_external_test_id.get(),
+                testCaseHash=current_test_case_hash.get(),
+                event=traced_event.to_json(),
+            ),
+            timeout=self._timeout_seconds,
+        )
+        req.raise_for_status()
+
+    def merge_properties(
```
Comment (nicolewhite) [inline autoblocks/_impl/tracer.py]: can you add a leading underscore since this is an internal function

Comment (adamnolte) [reply autoblocks/_impl/tracer.py]: ya i already did if you refresh 😄

File: autoblocks/_impl/testing/run.py (hunk 1 above)
Comment (adamnolte) [reply autoblocks/_impl/testing/run.py:148]: hmm ya good call, let me move it up a level

Comment (nicolewhite) [reply autoblocks/_impl/testing/run.py]: i don't think it matters too much since it's unnecessary like you said, was just a thought

Comment (adamnolte) [reply autoblocks/_impl/testing/run.py]: ya it doesn't matter for this use case, but good hygeine in case we use more and people reference this, I did the same for the test id too.

File: autoblocks/_impl/testing/run.py [hunk 7]
```diff
@@ -234,7 +239,7 @@ async def async_run_test_suite(
     }
 
     await global_state.http_client().post(f"{cli()}/start", json=dict(testExternalId=test_id))
-
```
Comment (nicolewhite) [inline autoblocks/_impl/testing/run.py:237]: would be nice to keep the original spacing

File: autoblocks/_impl/testing/run.py [hunk 8]
```diff
@@ -126,6 +128,7 @@ async def run_test_case_unsafe(
     Its caller will catch and handle all exceptions.
     """
     async with test_case_semaphore_registry[test_id]:
+
```
Comment (nicolewhite) [inline autoblocks/_impl/testing/run.py:131]: same here would be nice to drop this added space and the removed one below

**Review by nicolewhite — DISMISSED** [review]

File: autoblocks/_impl/testing/run.py (hunk 1 above)
Comment (nicolewhite) [reply autoblocks/_impl/testing/run.py]: ok nice

**Review by nicolewhite — APPROVED** [review]
